enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -1601,44 +1601,82 @@ func (p *PebbleCache) FindMissing(ctx context.Context, resources []*rspb.Resourc
 	}
 	defer db.Close()
 
+	lockedResources, err := p.lockResourceKeys(ctx, resources)
+	if err != nil {
+		return nil, err
+	}
 	iter := db.NewIter(nil /*default iterOptions*/)
 	defer iter.Close()
 
 	var missing []*repb.Digest
-	for _, r := range resources {
-		err = p.findMissing(ctx, iter, r)
+	for _, lockedResource := range lockedResources {
+		err = p.findMissing(ctx, db, iter, lockedResource.key)
 		if err != nil {
-			missing = append(missing, r.GetDigest())
+			missing = append(missing, lockedResource.resource.GetDigest())
 		}
+		lockedResource.unlockFn()
 	}
 	return missing, nil
 }
 
-func (p *PebbleCache) findMissing(ctx context.Context, iter pebble.Iterator, r *rspb.ResourceName) error {
-	fileRecord, err := p.makeFileRecord(ctx, r)
-	if err != nil {
-		return err
-	}
-	key, err := p.fileStorer.PebbleKey(fileRecord)
-	if err != nil {
-		return err
+type resourceLockInfo struct {
+	key      filestore.PebbleKey
+	unlockFn func()
+	resource *rspb.ResourceName
+}
+
+func (p *PebbleCache) lockResourceKeys(ctx context.Context, resources []*rspb.ResourceName) ([]*resourceLockInfo, error) {
+	res := make([]*resourceLockInfo, 0, len(resources))
```

Comment:
the keys would have to be sorted to avoid deadlocks
you could have 2 requests, one for `a, b` and one for `b, a`. if both locked their first digest first, neither would be able to make progress.